### PR TITLE
feat(extra-natives-five): add SET_AUDIO_SUBMIX_OUTPUT_VOLUMES native

### PR DESCRIPTION
**Pull request by manzarek123**

### CODE REVIEW

File: code/components/extra-natives-five/src/NuiAudioSink.cpp [hunk 1]
```diff
@@ -2038,6 +2038,34 @@ static InitFunction initFunction([]()
...
+
+	fx::ScriptEngine::RegisterNativeHandler("SET_AUDIO_SUBMIX_OUTPUT_VOLUMES", [](fx::ScriptContext& ctx)
+	{
+		int sourceIdx = ctx.GetArgument<int>(0);
+		uint32_t effectIdx = ctx.GetArgument<uint32_t>(1);
+		float leftVolume = ctx.GetArgument<float>(2);
+		float rightVolume = ctx.GetArgument<float>(3);
+		float rearLeftVolume = ctx.GetArgument<float>(4);
+		float rearRightVolume = ctx.GetArgument<float>(5);
+		float centerVolume = ctx.GetArgument<float>(6);
```
Comment (blattersturm) [inline code/components/extra-natives-five/src/NuiAudioSink.cpp:2051]: Have you confirmed the order in this case?

File: ext/native-decls/SetAudioSubmixOutputVolumes.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+```c
+void SET_AUDIO_SUBMIX_OUTPUT_VOLUMES(int submixId, int effectSlot, float leftVolume, float rightVolume, float rearLeftVolume, float rearRightVolume, float centerVolume, float lfeVolume);
+```
+
+Sets the volumes for the sound channels in a submix effect.
+Values can be between 0.0 and 1.0.
+
+## Parameters
+* **submixId**: The submix.
+* **effectSlot**: The effect slot for the submix. It is expected that the effect is set in this slot beforehand.
```
Comment (blattersturm) [inline ext/native-decls/SetAudioSubmixOutputVolumes.md:17]: Is it an effect slot still when it's an output slot rather?

File: code/components/extra-natives-five/src/NuiAudioSink.cpp (hunk 1 above)
Comment (manzarek123) [reply code/components/extra-natives-five/src/NuiAudioSink.cpp:2051]: I made some more tests with a 5.1 audio cable.
1 is left and seems to range from 0.0 to 1.0
2 is right and seems to range from 0.0 to 1.0
3 is surround left and seems to range from -1.0 to 1.0
4 is surround right and seems to range from -1.0 to 1.0
5 is unknown but not used in voice chat
6 is unknown but not used in voice chat

5 or 6 is center, other one must be lfe but cannot say for sure...

The game is not using center speaker (except for some menu sound and some minor stuff) and lfe also seems to not work according to reports online.

I made a video you can see the volume changes in the audio cable manager.
https://www.youtube.com/watch?v=AVf5NKslZR8

File: ext/native-decls/SetAudioSubmixOutputVolumes.md (hunk 2 above)
Comment (manzarek123) [reply ext/native-decls/SetAudioSubmixOutputVolumes.md:17]: And yes indeed, will fix and document this tomorrow

File: ext/native-decls/SetAudioSubmixOutputVolumes.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+void SET_AUDIO_SUBMIX_OUTPUT_VOLUMES(int submixId, int outputSubmixId, float frontLeftVolume, float frontRightVolume, float rearLeftVolume, float rearRightVolume, float channel5Volume, float channel6Volume);
+```
+
+Sets the volumes for the sound channels in a submix effect.
+Values can be between 0.0 and 1.0.
+Channel 5 and channel 6 are not used in voice chat but are believed to be center and LFE channels.
+
+## Parameters
+* **submixId**: The submix.
+* **outputSubmixId**: The output submix. Use 0 for the master game submix.
```
Comment (blattersturm) [inline ext/native-decls/SetAudioSubmixOutputVolumes.md:18]: Is this actually correct, and is it not an output slot as I suspected in initial documentation? (i.e. 0 will be the first AddOutput call, 1 the second, etc.)

Comment (manzarek123) [reply ext/native-decls/SetAudioSubmixOutputVolumes.md:18]: Yes I misunderstood again, this is exactly as you are saying

Full repro code I used to validate this:

local radioSubmix = CreateAudioSubmix('Radio')
SetAudioSubmixEffectRadioFx(radioSubmix, 0)
SetAudioSubmixEffectParamInt(radioSubmix,   0, `enabled`,     1)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `freq_low`,    202.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `freq_hi`,     6000.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `rm_mod_freq`, 0.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `rm_mix`,      0.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `fudge`,       5.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `o_freq_lo`,   141.0)
SetAudioSubmixEffectParamFloat(radioSubmix, 0, `o_freq_hi`,   6000.0)
AddAudioSubmixOutput(radioSubmix, 0)

local dummySubmix = CreateAudioSubmix('Dummy')
SetAudioSubmixEffectRadioFx(dummySubmix, 0)
SetAudioSubmixEffectParamInt(dummySubmix,   0, `enabled`,     1)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `freq_low`,    202.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `freq_hi`,     6000.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `rm_mod_freq`, 0.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `rm_mix`,      0.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `fudge`,       50.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `o_freq_lo`,   141.0)
SetAudioSubmixEffectParamFloat(dummySubmix, 0, `o_freq_hi`,   6000.0)

local radioSubmix2 = CreateAudioSubmix('Radio2')
SetAudioSubmixEffectRadioFx(radioSubmix2, 0)
SetAudioSubmixEffectParamInt(radioSubmix2,   0, `enabled`,     1)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `freq_low`,    402.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `freq_hi`,     6000.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `rm_mod_freq`, 0.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `rm_mix`,      0.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `fudge`,       00.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `o_freq_lo`,   141.0)
SetAudioSubmixEffectParamFloat(radioSubmix2, 0, `o_freq_hi`,   6000.0)
AddAudioSubmixOutput(radioSubmix2, dummySubmix)
AddAudioSubmixOutput(radioSubmix2, radioSubmix)

MumbleSetSubmixForServerId(1, radioSubmix2)

--SetAudioSubmixOutputVolumes(radioSubmix2, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0) -- will not mute
--SetAudioSubmixOutputVolumes(radioSubmix2, 1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0) -- will mute

--SetAudioSubmixOutputVolumes(radioSubmix, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0) -- will mute
--SetAudioSubmixOutputVolumes(radioSubmix, 1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0) -- will not mute

**Pull request merged**